crates/astria-composer/src/searcher/collector.rs

```diff
@@ -0,0 +1,221 @@
+///
+/// Used to send new transactions to the searcher.
+pub(super) struct Transaction {
+    pub(super) chain_id: String,
+    pub(super) inner: ethers::types::Transaction,
+}
+
+#[derive(Debug)]
+pub(super) struct Collector {
+    // Chain ID to identify in the astria sequencer block which rollup a serialized sequencer
+    // action belongs to.
+    chain_id: String,
+    // The client for getting new pending transactions from an ethereum rollup.
+    client: EthClient,
+    // The channel on which the collector sends new txs to the searcher.
+    searcher_channel: Sender<Transaction>,
+    // The status of this collector.
+    status: watch::Sender<Status>,
+}
+
+#[derive(Debug)]
+pub(super) struct Status {
+    is_connected: bool,
+}
+
+impl Status {
+    fn new() -> Self {
+        Self {
+            is_connected: false,
+        }
+    }
+
+    pub(super) fn is_connected(&self) -> bool {
+        self.is_connected
+    }
+}
+
+impl Collector {
+    pub(super) async fn new(
+        chain_id: String,
+        url: String,
+        searcher_channel: Sender<Transaction>,
+    ) -> eyre::Result<Self> {
+        let client = EthClient::connect(&url)
+            .await
+            .wrap_err("failed connecting to eth")?;
+        let (status, _) = watch::channel(Status::new());
+        Ok(Self {
+            chain_id,
+            client,
+            searcher_channel,
+            status,
+        })
+    }
+
+    /// Subscribe to the collector's status.
+    pub(super) fn subscribe(&self) -> watch::Receiver<Status> {
+        self.status.subscribe()
+    }
+
+    #[instrument(skip_all, fields(chain_id = self.chain_id))]
+    pub(super) async fn run_until_stopped(self) -> eyre::Result<()> {
+        use std::time::Duration;
+
+        use ethers::providers::Middleware as _;
+        use futures::stream::StreamExt as _;
+        self.wait_for_evm(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed connecting ")?;
+
+        let Self {
+            chain_id,
+            client,
+            searcher_channel,
+            ..
+        } = self;
+
+        let mut tx_stream = client
+            .inner
+            .subscribe_full_pending_txs()
+            .await
+            .wrap_err("failed to subscribe eth client to full pending transactions")?;
+        while let Some(tx) = tx_stream.next().await {
+            match searcher_channel
+                .send_timeout(
+                    Transaction {
+                        chain_id: chain_id.clone(),
+                        inner: tx,
+                    },
+                    Duration::from_millis(500),
+                )
+                .await
+            {
+                Ok(()) => {}
+                Err(SendTimeoutError::Timeout(_tx)) => {
+                    // TODO(superfluffy): what parts of tx should we log? just its hash?
+                    warn!("timed out sending new transaction to searcher after 500ms; dropping tx");
```

Comment:
We should probably log the chainid + the hash. That will uniquely identify the tx.

Reply:
Logging the hash. `chain_id` was already logged as part of the instrumented scope.